br/cmd/br/cmd.go

```diff
@@ -162,6 +165,32 @@ func Init(cmd *cobra.Command) (err error) {
 		}
 		log.ReplaceGlobals(lg, p)
 		memory.InitMemoryHook()
+		if debug.SetMemoryLimit(-1) == math.MaxInt64 {
+			memtotal, e := memory.MemTotal()
+			if e != nil {
+				err = e
+				return
+			}
+			memused, e := memory.MemUsed()
+			if e != nil {
+				err = e
+				return
+			}
+			halfGB := uint64(512 * 1024 * 1024)
+			// 0     memused          memtotal-512MB     memtotal
+			// +--------+--------------------+--------------+
+			//                       br mem upper limit
+			memleft := memtotal - memused - halfGB
+			// 0       512MB               +inf
+			// +--------+-------------------->
+			//          ^____________________>
+			//             GOMEMLIMIT range
+			log.Info("calculate the rest memory", zap.Uint64("memtotal", memtotal), zap.Uint64("memused", memused))
+			if memleft > halfGB {
+				log.Info("set memory limit", zap.Uint64("limit", memleft))
+				debug.SetMemoryLimit(int64(memleft))
+			}
```

Comment:
if `memtotal - memused < halfGB` the subtraction in `memleft` will overflow causing BR to set GOMEMLIMIT to an absurdly large value.

Reply:
the first `halfGB` is aimed to keep the reserved memory to avoid OOM.
the second `halfGB` is aimed to give up use memory limit because the `memleft` may be too small for br. I am worried that there is a high probability that it triggers frequent GC and then oom.

Besides, the current method of calculating reserved memory is too coarse, I will modify it.